File: cost_tracker.py
"""Thread-safe LLM cost tracking with per-model pricing."""

import threading

# Pricing per 1M tokens: (input, cache_read_input, output)
# Anthropic cache writes: billed at 1.25× input (5-min TTL) or 2× input (1-hour TTL).
# OpenAI cache writes: free (automatic); only cache reads are discounted.
_PRICING = {
    # OpenAI
    "gpt-5-mini":    (0.25, 0.025,  2.00),
    "gpt-4.1-mini":  (0.40, 0.10,   1.60),
    # Anthropic — Claude 4.x
    "claude-haiku":  (1.00, 0.10,   5.00),   # Haiku 4.5
    "claude-sonnet": (3.00, 0.30,  15.00),   # Sonnet 4.6
    "claude-opus":   (5.00, 0.50,  25.00),   # Opus 4.6
}


def _lookup_pricing(model):
    """Return (input_price, cache_read_price, output_price) per 1M tokens for a model."""
    m = model.lower()
    # Sort by key length descending so the most-specific key wins.
    for key, prices in sorted(_PRICING.items(), key=lambda x: -len(x[0])):
        if key in m:
            return prices
    return (1.00, 0.10, 3.00)  # conservative fallback


class CostTracker:
    """Accumulates token usage across a pipeline session.

    Token categories tracked separately:
        input_tokens           — non-cached input tokens (billed at input_price)
        cache_creation_tokens  — tokens written to cache on Anthropic
                                 (billed at 1.25× input for 5-min TTL;
                                  OpenAI caching is free so this is always 0 there)
        cache_read_tokens      — tokens read from cache (billed at cache_read_price,
                                 ~0.1× input for both providers)
        output_tokens          — generated output tokens
    """

    def __init__(self):
        self._lock = threading.Lock()
        self._input_tokens = 0
        self._cache_creation_tokens = 0
        self._cache_read_tokens = 0
        self._output_tokens = 0

    def add_tokens(self, input_tokens, output_tokens,
                   cache_creation_tokens=0, cache_read_tokens=0):
        with self._lock:
            self._input_tokens += input_tokens
            self._cache_creation_tokens += cache_creation_tokens
            self._cache_read_tokens += cache_read_tokens
            self._output_tokens += output_tokens

    def reset(self):
        with self._lock:
            self._input_tokens = 0
            self._cache_creation_tokens = 0
            self._cache_read_tokens = 0
            self._output_tokens = 0

    def get_session_cost(self, model):
        inp_price, cache_read_price, out_price = _lookup_pricing(model)
        cache_write_price = inp_price * 1.25  # conservative: 5-min TTL multiplier
        with self._lock:
            return (
                self._input_tokens            * inp_price
                + self._cache_creation_tokens * cache_write_price
                + self._cache_read_tokens     * cache_read_price
                + self._output_tokens         * out_price
            ) / 1_000_000

    def get_tokens(self):
        """Return (input, cache_creation, cache_read, output) token counts."""
        with self._lock:
            return (
                self._input_tokens,
                self._cache_creation_tokens,
                self._cache_read_tokens,
                self._output_tokens,
            )

    @staticmethod
    def estimate_summarization_cost(article_count, model):
        """Estimate cost for summarizing a batch of articles with prompt caching.

        The expanded SUMMARY_PROMPT (~4,500 tokens) is written to cache on the
        first article and read from cache on all subsequent articles in the batch.
        """
        if article_count == 0:
            return 0.0
        inp_price, cache_read_price, out_price = _lookup_pricing(model)
        cache_write_price = inp_price * 1.25
        system_tokens = 4500   # expanded SUMMARY_PROMPT with MITRE reference
        user_tokens   = 3000   # typical article content
        out_tokens    = 500    # typical summary output

        first = (system_tokens * cache_write_price
                 + user_tokens * inp_price
                 + out_tokens  * out_price) / 1_000_000
        rest  = max(0, article_count - 1) * (
                 system_tokens * cache_read_price
                 + user_tokens * inp_price
                 + out_tokens  * out_price) / 1_000_000
        return (first + rest) * 1.5  # 1.5× buffer for variance


# Singleton instance used by the whole application
cost_tracker = CostTracker()

File: android/app/src/main/java/com/threatloom/app/domain/service/CostTracker.kt
package com.threatloom.app.domain.service

import java.util.concurrent.atomic.AtomicInteger
import javax.inject.Inject
import javax.inject.Singleton

data class CostSnapshot(
    val input: Int = 0,
    val output: Int = 0,
    val cacheWrite: Int = 0,
    val cacheRead: Int = 0
)

@Singleton
class CostTracker @Inject constructor() {

    private val inputTokens = AtomicInteger(0)
    private val outputTokens = AtomicInteger(0)
    private val cacheWriteTokens = AtomicInteger(0)
    private val cacheReadTokens = AtomicInteger(0)

    fun addTokens(input: Int, output: Int) {
        inputTokens.addAndGet(input)
        outputTokens.addAndGet(output)
    }

    fun addUsage(input: Int, output: Int, cacheWrite: Int = 0, cacheRead: Int = 0) {
        inputTokens.addAndGet(input)
        outputTokens.addAndGet(output)
        cacheWriteTokens.addAndGet(cacheWrite)
        cacheReadTokens.addAndGet(cacheRead)
    }

    fun reset() {
        inputTokens.set(0)
        outputTokens.set(0)
        cacheWriteTokens.set(0)
        cacheReadTokens.set(0)
    }

    fun getSnapshot(): CostSnapshot = CostSnapshot(
        inputTokens.get(), outputTokens.get(),
        cacheWriteTokens.get(), cacheReadTokens.get()
    )

    fun getTokens(): Pair<Int, Int> = inputTokens.get() to outputTokens.get()

    fun getSessionCost(model: String): Double {
        val (inputPrice, cacheReadPrice, outputPrice) = pricingPer1M(model)
        val cacheWritePrice = cacheWritePrice(model, inputPrice)
        return (inputTokens.get() * inputPrice
                + outputTokens.get() * outputPrice
                + cacheWriteTokens.get() * cacheWritePrice
                + cacheReadTokens.get() * cacheReadPrice) / 1_000_000.0
    }

    fun estimateSummarizationCost(articleCount: Int, model: String): Double {
        if (articleCount <= 0) return 0.0
        val (inputPrice, cacheReadPrice, outputPrice) = pricingPer1M(model)
        // Conservative write rate for the first-article system prompt (matches the web
        // estimate_summarization_cost; kept provider-agnostic so both apps agree).
        val cacheWritePrice = inputPrice * 1.25
        // The large summary system prompt is written to cache on the first article
        // and read from cache on every subsequent article in the batch.
        val systemTokens = 4500.0   // expanded SUMMARY_PROMPT with extraction example
        val userTokens = 3000.0     // typical article content
        val outTokens = 500.0       // typical summary output
        val first = (systemTokens * cacheWritePrice
                + userTokens * inputPrice
                + outTokens * outputPrice) / 1_000_000.0
        val rest = maxOf(0, articleCount - 1) * (systemTokens * cacheReadPrice
                + userTokens * inputPrice
                + outTokens * outputPrice) / 1_000_000.0
        return (first + rest) * 1.5  // 1.5x buffer for variance
    }

    fun estimateTrendCost(
        articles: List<com.threatloom.app.domain.model.ArticleWithSummary>,
        model: String
    ): Triple<Double, Int, Int> {
        val groups = articles.groupBy { a ->
            val parts = (a.publishedDate ?: "").split("-")
            val year = parts.getOrNull(0)?.toIntOrNull() ?: 0
            val month = parts.getOrNull(1)?.toIntOrNull() ?: 0
            year to ((month - 1) / 3 + 1)
        }.filterKeys { (y, _) -> y > 0 }
        val nQ = groups.size
        val nY = groups.keys.map { it.first }.toSet().size
        val nBatches = groups.values.sumOf { maxOf(0, (it.size - 1) / 50) }
        val (inp, _, out) = pricingPer1M(model)
        val cost = ((nQ * 3000 + nY * 8000 + nBatches * 15000) * inp +
                    (nQ * 2500 + nY * 3000 + nBatches * 1500) * out) / 1_000_000.0
        return Triple(cost, nQ, nY)
    }

    fun estimateInsightCost(articleCount: Int, model: String): Double {
        val (inp, _, out) = pricingPer1M(model)
        val estInput = minOf(articleCount * 200, 5000) + 200
        return (estInput * inp + 2000.0 * out) / 1_000_000.0
    }

    fun deltaCost(before: CostSnapshot, after: CostSnapshot, model: String): Double {
        val (inputPrice, cacheReadPrice, outputPrice) = pricingPer1M(model)
        val cacheWritePrice = cacheWritePrice(model, inputPrice)
        return ((after.input - before.input) * inputPrice
                + (after.output - before.output) * outputPrice
                + (after.cacheWrite - before.cacheWrite) * cacheWritePrice
                + (after.cacheRead - before.cacheRead) * cacheReadPrice) / 1_000_000.0
    }

    fun deltaCost(before: Pair<Int, Int>, after: Pair<Int, Int>, model: String): Double {
        val (inp, _, out) = pricingPer1M(model)
        return ((after.first - before.first) * inp + (after.second - before.second) * out) / 1_000_000.0
    }

    /** Anthropic charges 1.25x input for 5-min cache writes; OpenAI cache writes are free. */
    private fun cacheWritePrice(model: String, inputPrice: Double): Double {
        return if ("claude" in model.lowercase()) inputPrice * 1.25 else 0.0
    }

    // Pricing per 1M tokens: (input, cacheRead, output). Only the models offered in settings.
    private fun pricingPer1M(model: String): Triple<Double, Double, Double> {
        val m = model.lowercase()
        return when {
            "gpt-5-mini" in m -> Triple(0.25, 0.025, 2.00)
            "gpt-5.4-nano" in m -> Triple(0.20, 0.020, 1.25)
            "claude-haiku" in m -> Triple(1.00, 0.10, 5.00)
            "claude-sonnet" in m -> Triple(3.00, 0.30, 15.00)
            else -> Triple(1.00, 0.10, 3.00)
        }
    }
}
